docs/guides/environment_variables.md
---
id: environment-variables
title: Environment Variables
---

The following is a list of the environment variables used by Apify SDK that are available to the user.
The SDK is capable of running without any env vars present, but certain features will only become available
after env vars are properly set. You can use [Apify CLI](https://github.com/apify/apify-cli)
to set the env vars for you. [Apify platform](../guides/apify-platform) also sets the variables automatically.

## Important env vars

The following environment variables have large impact on the way Apify SDK works and its behavior
can be changed significantly by setting or unsetting them.

### `APIFY_LOCAL_STORAGE_DIR`

Defines the path to a local directory where [`KeyValueStore`](../api/key-value-store),
[`Dataset`](../api/dataset), and [`RequestQueue`](../api/request-queue) store their data.
Typically it is set to `./apify_storage`. If omitted, you should define the [`APIFY_TOKEN`](#apify_token)
environment variable instead.

### `APIFY_TOKEN`

The API token for your Apify account. It is used to access the Apify API, e.g. to access cloud storage
or to run an actor on the Apify platform. You can find your API token on the
[Account - Integrations](https://console.apify.com/account#/integrations) page. If omitted,
you should define the `APIFY_LOCAL_STORAGE_DIR` environment variable instead.

### Combinations of `APIFY_LOCAL_STORAGE_DIR` and `APIFY_TOKEN`

By combining the env vars in various ways, you can greatly influence the behavior of Apify SDK.

| Env Vars                                    | API | Storages       |
| ------------------------------------------- | --- | -------------- |
|  none OR `APIFY_LOCAL_STORAGE_DIR`          | no  | local          |
| `APIFY_TOKEN`                               | yes | Apify platform |
| `APIFY_TOKEN` AND `APIFY_LOCAL_STORAGE_DIR` | yes | local+platform |

When using both `APIFY_TOKEN` and `APIFY_LOCAL_STORAGE_DIR`, you can use all the Apify platform
features and your data will be stored locally by default. If you want to access platform storages,
you can use the `{ forceCloud: true }` option in their respective functions.

```js
const localDataset = await Apify.openDataset('my-local-data');
const remoteDataset = await Apify.openDataset('my-remote-data', { forceCloud: true });
```

## Convenience env vars

The next group includes env vars that can help achieve certain goals without having to change
your code, such as temporarily switching log level to DEBUG.

### `APIFY_HEADLESS`

If set to `1`, web browsers launched by Apify SDK will run in the headless mode. You can still override
this setting in the code, e.g. by passing the `headless: true` option to the
[`Apify.launchPuppeteer()`](../api/apify#launchpuppeteer) function. But having this setting
in an environment variable allows you to develop the crawler locally in headful mode to simplify the debugging,
and only run the crawler in headless mode once you deploy it to the Apify platform. By default, the browsers
are launched in headful mode, i.e. with windows.

### `APIFY_LOG_LEVEL`

Specifies the minimum log level, which can be one of the following values (in order of severity):
`DEBUG`, `INFO`, `WARNING` and `ERROR`. By default, the log level is set to `INFO`,
which means that `DEBUG` messages are not printed to console. See the [`utils.log`](../api/log)
namespace for logging utilities.

### `APIFY_MEMORY_MBYTES`

Sets the amount of system memory in megabytes to be used by the [`AutoscaledPool`](../api/autoscaled-pool).
It is used to limit the number of concurrently running tasks. By default, the max amount of memory
to be used is set to one quarter of total system memory, i. e. on a system with 8192 MB of memory,
the autoscaling feature will only use up to 2048 MB of memory.

### `APIFY_PROXY_PASSWORD`

Optional password to [Apify Proxy](https://docs.apify.com/proxy) for IP address rotation.
If you have an Apify Account, you can find the password on the [Proxy page](https://console.apify.com/proxy)
in the Apify Console. The password is automatically inferred using the `APIFY_TOKEN` env var,
so in most cases, you don't need to touch it. You should use it when, for some reason,
you need access to Apify Proxy, but no access to Apify API, or when you need access to
proxy from a different account than your token represents.


v2-website/versioned_docs/version-1.3.1/api/Configuration.md
---
id: configuration
title: Configuration
original_id: configuration
---

<a name="configuration"></a>

`Configuration` is a value object holding the SDK configuration. We can use it in two ways:

1. When using `Apify` class, we can get the instance configuration via `sdk.config`

```js
const { Apify } = require('apify');

const sdk = new Apify({ token: '123' });
console.log(sdk.config.get('token')); // '123'
```

2. To get the global configuration (singleton instance). It will respect the environment variables.

```js
console.log(Configuration.getGlobalConfig().get('token')); // returns the token from APIFY_TOKEN env var
```

## Supported Configuration Options

| Key                          | Environment Variable                  | Default Value       |
| ---------------------------- | ------------------------------------- | ------------------- |
| `defaultDatasetId`           | `APIFY_DEFAULT_DATASET_ID`            | `'default'`         |
| `defaultKeyValueStoreId`     | `APIFY_DEFAULT_KEY_VALUE_STORE_ID`    | `'default'`         |
| `defaultRequestQueueId`      | `APIFY_DEFAULT_REQUEST_QUEUE_ID`      | `'default'`         |
| `localStorageDir`            | `APIFY_LOCAL_STORAGE_DIR`             | `'./apify_storage'` |
| `localStorageEnableWalMode`  | `APIFY_LOCAL_STORAGE_ENABLE_WAL_MODE` | `true`              |
| `persistStateIntervalMillis` | `APIFY_PERSIST_STATE_INTERVAL_MILLIS` | `60e3`              |
| `token`                      | `APIFY_TOKEN`                         | -                   |

## Advanced Configuration Options

| Key                         | Environment Variable                 | Default Value              |
| --------------------------- | ------------------------------------ | -------------------------- |
| `actorEventsWsUrl`          | `APIFY_ACTOR_EVENTS_WS_URL`          | -                          |
| `actorId`                   | `APIFY_ACTOR_ID`                     | -                          |
| `actorRunId`                | `APIFY_ACTOR_RUN_ID`                 | -                          |
| `actorTaskId`               | `APIFY_ACTOR_TASK_ID`                | -                          |
| `apiBaseUrl`                | `APIFY_API_BASE_URL`                 | `'https://api.apify.com'`  |
| `containerPort`             | `APIFY_CONTAINER_PORT`               | `4321`                     |
| `containerUrl`              | `APIFY_CONTAINER_URL`                | `'http://localhost:4321'`  |
| `inputKey`                  | `APIFY_INPUT_KEY`                    | `'INPUT'`                  |
| `isAtHome`                  | `APIFY_IS_AT_HOME`                   | -                          |
| `metamorphAfterSleepMillis` | `APIFY_METAMORPH_AFTER_SLEEP_MILLIS` | `300e3`                    |
| `proxyHostname`             | `APIFY_PROXY_HOSTNAME`               | `'proxy.apify.com'`        |
| `proxyPassword`             | `APIFY_PROXY_PASSWORD`               | -                          |
| `proxyPort`                 | `APIFY_PROXY_PORT`                   | `8000`                     |
| `proxyStatusUrl`            | `APIFY_PROXY_STATUS_URL`             | `'http://proxy.apify.com'` |
| `userId`                    | `APIFY_USER_ID`                      | -                          |

## Not Supported environment variables

-   `MEMORY_MBYTES`
-   `HEADLESS`
-   `XVFB`
-   `CHROME_EXECUTABLE_PATH`

---

<a name="exports.configuration"></a>

## `new Configuration(options)`

Creates new `Configuration` instance with provided options. Env vars will have precedence over those.

**Parameters**:

-   **`options`**: `Record<string, (number|string|boolean)>`

---

<a name="get"></a>

## `configuration.get(key, [defaultValue])`

Returns configured value. First checks the environment variables, then provided configuration, fallbacks to the `defaultValue` argument if provided,
otherwise uses the default value as described in the above section.

**Parameters**:

-   **`key`**: `string`
-   **`[defaultValue]`**: `string` | `number` | `boolean`

**Returns**:

`string` \| `number` \| `boolean`

---

<a name="set"></a>

## `configuration.set(key, [value])`

Sets value for given option. Only affects this `Configuration` instance, the value will not be propagated down to the env var. To reset a value, we
can omit the `value` argument or pass `undefined` there.

**Parameters**:

-   **`key`**: `string`
-   **`[value]`**: `string` | `number` | `boolean`

---

<a name="getclient"></a>

## `configuration.getClient([options])`

Returns cached instance of [`ApifyClient`](../api/apify) using options as defined in the environment variables or in this
[`Configuration`](../api/configuration) instance. Only first call of this method will create the client, following calls will return the same client
instance.

Caching works based on the API URL and token, so calling this method with different options will return multiple instances, one for each variant of
the options.

**Internal**:  
**Parameters**:

-   **`[options]`**: `object`
    -   **`[token]`**: `string`
    -   **`[maxRetries]`**: `string`
    -   **`[minDelayBetweenRetriesMillis]`**: `string`

**Returns**:

[`ApifyClient`](../api/apify)

---

<a name="getstoragelocal"></a>

## `configuration.getStorageLocal([options])`

Returns cached instance of [`ApifyStorageLocal`](../api/apify) using options as defined in the environment variables or in this
[`Configuration`](../api/configuration) instance. Only first call of this method will create the client, following calls will return the same client
instance.

Caching works based on the `storageDir` option, so calling this method with different `storageDir` will return multiple instances, one for each
directory.

**Internal**:  
**Parameters**:

-   **`[options]`**: `object`
    -   **`[storageDir]`**: `string`
    -   **`[enableWalMode]`**: `boolean` <code> = true</code>

**Returns**:

[`ApifyStorageLocal`](../api/apify)

---

<a name="createclient"></a>

## `configuration.createClient([options])`

Creates an instance of ApifyClient using options as defined in the environment variables or in this `Configuration` instance.

**Internal**:  
**Parameters**:

-   **`[options]`**: `object`
    -   **`[token]`**: `string`
    -   **`[maxRetries]`**: `string`
    -   **`[minDelayBetweenRetriesMillis]`**: `string`

**Returns**:

[`ApifyClient`](../api/apify)

---

<a name="createstoragelocal"></a>

## `configuration.createStorageLocal([options])`

Creates an instance of ApifyStorageLocal using options as defined in the environment variables or in this `Configuration` instance.

**Internal**:  
**Parameters**:

-   **`[options]`**: `object`
    -   **`[storageDir]`**: `string`
    -   **`[enableWalMode]`**: `boolean` <code> = true</code>

**Returns**:

[`ApifyStorageLocal`](../api/apify)

---

<a name="getglobalconfig"></a>

## `Configuration.getGlobalConfig()`

Returns the global configuration instance. It will respect the environment variables. As opposed to this method, we can also get the SDK instance
configuration via `sdk.config` property.

**Returns**:

[`Configuration`](../api/configuration)

---


src/configuration.ts
import { ENV_VARS, LOCAL_ENV_VARS } from '@apify/consts';
import { join } from 'path';
import { ApifyStorageLocal } from '@apify/storage-local';
import { ApifyClient } from 'apify-client';
import log from './utils_log';

export interface ConfigurationOptions {
    token?: string;
    localStorageDir?: string;
    localStorageEnableWalMode?: boolean;
    defaultDatasetId?: string;
    defaultKeyValueStoreId?: string;
    defaultRequestQueueId?: string;
    metamorphAfterSleepMillis?: number;
    persistStateIntervalMillis?: number;
    actorEventsWsUrl?: string;
    inputKey?: string;
    actorId?: string;
    apiBaseUrl?: string;
    isAtHome?: boolean;
    actorRunId?: string;
    actorTaskId?: string;
    containerPort?: number;
    containerUrl?: string;
    userId?: string;
    proxyHostname?: string;
    proxyPassword?: string;
    proxyStatusUrl?: string;
    proxyPort?: number;
}

/**
 * `Configuration` is a value object holding the SDK configuration. We can use it in two ways:
 *
 * 1. When using `Apify` class, we can get the instance configuration via `sdk.config`
 *   ```js
 *   const { Apify } = require('apify');
 *
 *   const sdk = new Apify({ token: '123' });
 *   console.log(sdk.config.get('token')); // '123'
 *   ```
 * 2. To get the global configuration (singleton instance). It will respect the environment variables.
 *   ```js
 *   // returns the token from APIFY_TOKEN env var
 *   console.log(Configuration.getGlobalConfig().get('token'));
 *   ```
 *
 * ## Supported Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `defaultDatasetId` | `APIFY_DEFAULT_DATASET_ID` | `'default'`
 * `defaultKeyValueStoreId` | `APIFY_DEFAULT_KEY_VALUE_STORE_ID` | `'default'`
 * `defaultRequestQueueId` | `APIFY_DEFAULT_REQUEST_QUEUE_ID` | `'default'`
 * `localStorageDir` | `APIFY_LOCAL_STORAGE_DIR` | `'./apify_storage'`
 * `localStorageEnableWalMode` | `APIFY_LOCAL_STORAGE_ENABLE_WAL_MODE` | `true`
 * `persistStateIntervalMillis` | `APIFY_PERSIST_STATE_INTERVAL_MILLIS` | `60e3`
 * `token` | `APIFY_TOKEN` | -
 *
 * ## Advanced Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `actorEventsWsUrl` | `APIFY_ACTOR_EVENTS_WS_URL` | -
 * `actorId` | `APIFY_ACTOR_ID` | -
 * `actorRunId` | `APIFY_ACTOR_RUN_ID` | -
 * `actorTaskId` | `APIFY_ACTOR_TASK_ID` | -
 * `apiBaseUrl` | `APIFY_API_BASE_URL` | `'https://api.apify.com'`
 * `containerPort` | `APIFY_CONTAINER_PORT` | `4321`
 * `containerUrl` | `APIFY_CONTAINER_URL` | `'http://localhost:4321'`
 * `inputKey` | `APIFY_INPUT_KEY` | `'INPUT'`
 * `isAtHome` | `APIFY_IS_AT_HOME` | -
 * `metamorphAfterSleepMillis` | `APIFY_METAMORPH_AFTER_SLEEP_MILLIS` | `300e3`
 * `proxyHostname` | `APIFY_PROXY_HOSTNAME` | `'proxy.apify.com'`
 * `proxyPassword` | `APIFY_PROXY_PASSWORD` | -
 * `proxyPort` | `APIFY_PROXY_PORT` | `8000`
 * `proxyStatusUrl` | `APIFY_PROXY_STATUS_URL` | `'http://proxy.apify.com'`
 * `userId` | `APIFY_USER_ID` | -
 *
 * ## Not Supported environment variables
 *
 * - `MEMORY_MBYTES`
 * - `HEADLESS`
 * - `XVFB`
 * - `CHROME_EXECUTABLE_PATH`
 */
export class Configuration {
    /**
     * Maps environment variables to config keys (e.g. `APIFY_PROXY_PORT` to `proxyPort`)
     */
    private static ENV_MAP = {
        APIFY_TOKEN: 'token',
        APIFY_LOCAL_STORAGE_DIR: 'localStorageDir',
        APIFY_LOCAL_STORAGE_ENABLE_WAL_MODE: 'localStorageEnableWalMode',
        APIFY_DEFAULT_DATASET_ID: 'defaultDatasetId',
        APIFY_DEFAULT_KEY_VALUE_STORE_ID: 'defaultKeyValueStoreId',
        APIFY_DEFAULT_REQUEST_QUEUE_ID: 'defaultRequestQueueId',
        APIFY_METAMORPH_AFTER_SLEEP_MILLIS: 'metamorphAfterSleepMillis',
        APIFY_PERSIST_STATE_INTERVAL_MILLIS: 'persistStateIntervalMillis',
        APIFY_TEST_PERSIST_INTERVAL_MILLIS: 'persistStateIntervalMillis', // for BC, seems to be unused
        APIFY_ACTOR_EVENTS_WS_URL: 'actorEventsWsUrl',
        APIFY_INPUT_KEY: 'inputKey',
        APIFY_ACTOR_ID: 'actorId',
        APIFY_API_BASE_URL: 'apiBaseUrl',
        APIFY_IS_AT_HOME: 'isAtHome',
        APIFY_ACTOR_RUN_ID: 'actorRunId',
        APIFY_ACTOR_TASK_ID: 'actorTaskId',
        APIFY_CONTAINER_PORT: 'containerPort',
        APIFY_CONTAINER_URL: 'containerUrl',
        APIFY_USER_ID: 'userId',
        APIFY_PROXY_HOSTNAME: 'proxyHostname',
        APIFY_PROXY_PASSWORD: 'proxyPassword',
        APIFY_PROXY_STATUS_URL: 'proxyStatusUrl',
        APIFY_PROXY_PORT: 'proxyPort',

        // not supported, use env vars directly:
        // APIFY_MEMORY_MBYTES: 'memoryMbytes',
        // APIFY_HEADLESS: 'headless',
        // APIFY_XVFB: 'xvfb',
        // APIFY_CHROME_EXECUTABLE_PATH: 'chromeExecutablePath',
    };

    /**
     * maps config keys to environment variables (e.g. `proxyPort` to `APIFY_PROXY_PORT`)
     */
    private static ENV_MAP_REVERSED = Object.entries(Configuration.ENV_MAP).reduce((obj, [key, value]) => {
        obj[value] = key;
        return obj;
    }, {} as Record<string, string>);

    private static BOOLEAN_VARS = ['localStorageEnableWalMode'];

    private static INTEGER_VARS = ['proxyPort', 'memoryMbytes', 'containerPort'];

    private static DEFAULTS = {
        defaultKeyValueStoreId: LOCAL_ENV_VARS[ENV_VARS.DEFAULT_KEY_VALUE_STORE_ID],
        defaultDatasetId: LOCAL_ENV_VARS[ENV_VARS.DEFAULT_DATASET_ID],
        defaultRequestQueueId: LOCAL_ENV_VARS[ENV_VARS.DEFAULT_REQUEST_QUEUE_ID],
        inputKey: 'INPUT',
        apiBaseUrl: 'https://api.apify.com',
        proxyStatusUrl: 'http://proxy.apify.com',
        proxyHostname: LOCAL_ENV_VARS[ENV_VARS.PROXY_HOSTNAME],
        proxyPort: +LOCAL_ENV_VARS[ENV_VARS.PROXY_PORT],
        containerPort: +LOCAL_ENV_VARS[ENV_VARS.CONTAINER_PORT],
        containerUrl: LOCAL_ENV_VARS[ENV_VARS.CONTAINER_URL],
        metamorphAfterSleepMillis: 300e3,
        persistStateIntervalMillis: 60e3, // This value is mentioned in jsdoc in `events.js`, if you update it here, update it there too.
        localStorageEnableWalMode: true,
    };

    private options: Map<keyof ConfigurationOptions, ConfigurationOptions[keyof ConfigurationOptions]>;
    private services = new Map<string, unknown>();
    private static globalConfig?: Configuration;

    /**
     * Creates new `Configuration` instance with provided options. Env vars will have precedence over those.
     */
    constructor(options: ConfigurationOptions = {}) {
        this.options = new Map(Object.entries(options)) as any;

        if (!this.get('localStorageDir') && !this.get('token')) {
            const dir = join(process.cwd(), './apify_storage');
            this.set('localStorageDir', dir);
            // eslint-disable-next-line max-len
            log.warning(`Neither ${ENV_VARS.LOCAL_STORAGE_DIR} nor ${ENV_VARS.TOKEN} environment variable is set, defaulting to ${ENV_VARS.LOCAL_STORAGE_DIR}="${dir}"`);
        }
    }

    /**
     * Returns configured value. First checks the environment variables, then provided configuration,
     * fallbacks to the `defaultValue` argument if provided, otherwise uses the default value as described
     * in the above section.
     */
    get<T extends keyof ConfigurationOptions, U extends ConfigurationOptions[T]>(key: T, defaultValue?: U): U {
        // prefer env vars
        const envKey = Configuration.ENV_MAP_REVERSED[key] ?? key;
        const envValue = process.env[envKey];

        if (envValue != null) {
            return this._castEnvValue(key, envValue);
        }

        // check instance level options
        if (this.options.has(key)) {
            return this.options.get(key) as U;
        }

        // fallback to defaults
        return (defaultValue ?? Configuration.DEFAULTS[key as string] ?? envValue) as U;
    }

    /**
     * @param {string} key
     * @param {number | string | boolean} value
     * @return {boolean}
     * @private
     */
    _castEnvValue(key, value) {
        if (Configuration.INTEGER_VARS.includes(key)) {
            return +value;
        }

        if (Configuration.BOOLEAN_VARS.includes(key)) {
            // 0, false and empty string are considered falsy values
            return !['0', 'false', ''].includes(value.toLowerCase());
        }

        return value;
    }

    /**
     * Sets value for given option. Only affects this `Configuration` instance, the value will not be propagated down to the env var.
     * To reset a value, we can omit the `value` argument or pass `undefined` there.
     *
     * @param {string} key
     * @param {string | number | boolean} [value]
     */
    set(key, value) {
        this.options.set(key, value);
    }

    /**
     * Returns cached instance of {@link ApifyClient} using options as defined in the environment variables or in
     * this {@link Configuration} instance. Only first call of this method will create the client, following calls will
     * return the same client instance.
     *
     * Caching works based on the API URL and token, so calling this method with different options will return
     * multiple instances, one for each variant of the options.
     * @internal
     */
    getClient(options: { token?: string; maxRetries?: string; minDelayBetweenRetriesMillis?: string; baseUrl?: string } = {}): ApifyClient {
        const baseUrl = options.baseUrl ?? this.get('apiBaseUrl');
        const token = options.token ?? this.get('token');
        const cacheKey = `${baseUrl}~${token}`;

        return this._getService('ApifyClient', () => this.createClient(options), cacheKey);
    }

    /**
     * Returns cached instance of {@link ApifyStorageLocal} using options as defined in the environment variables or in
     * this {@link Configuration} instance. Only first call of this method will create the client, following calls will return
     * the same client instance.
     *
     * Caching works based on the `storageDir` option, so calling this method with different `storageDir` will return
     * multiple instances, one for each directory.
     *
     * @param [options.storageDir]
     * @param [options.enableWalMode=true]
     * @return {ApifyStorageLocal}
     * @internal
     */
    getStorageLocal(options: { storageDir?: string; enableWalMode?: boolean } = {}) {
        const cacheKey = options.storageDir ?? this.get('localStorageDir');
        return this._getService('ApifyStorageLocal', () => this.createStorageLocal(options), cacheKey);
    }

    /**
     * Returns cached (singleton) instance of a service by its name. If the service does not exist yet, it will be created
     * via the `createCallback`. To have multiple instances of one service, we can use unique values in the `cacheKey`.
     *
     * @param {string} name
     * @param {Function} createCallback
     * @param {string} [cacheKey]
     * @return {unknown}
     * @private
     */
    _getService<T = unknown>(name: string, createCallback: () => T, cacheKey = name): T {
        cacheKey = `${name}~${cacheKey}`;

        if (!this.services.has(cacheKey)) {
            this.services.set(cacheKey, createCallback());
        }

        return this.services.get(cacheKey) as T;
    }

    /**
     * Creates an instance of ApifyClient using options as defined in the environment variables or in this `Configuration` instance.
     *
     * @param {object} [options]
     * @param {string} [options.token]
     * @param {string} [options.maxRetries]
     * @param {string} [options.minDelayBetweenRetriesMillis]
     * @return {ApifyClient}
     * @internal
     */
    createClient(options = {}) {
        return new ApifyClient({
            baseUrl: this.get('apiBaseUrl'),
            token: this.get('token'),
            ...options, // allow overriding the instance configuration
        });
    }

    /**
     * Creates an instance of ApifyStorageLocal using options as defined in the environment variables or in this `Configuration` instance.
     * @internal
     */
    createStorageLocal(options: { storageDir?: string; enableWalMode?: boolean } = {}): ApifyStorageLocal {
        const storageDir = options.storageDir ?? this.get('localStorageDir');
        const enableWalMode = options.enableWalMode ?? this.get('localStorageEnableWalMode');
        const storage = new ApifyStorageLocal({ ...options, storageDir, enableWalMode });

        process.on('exit', () => {
            // TODO this is not public API, need to update
            // storage local with some teardown
            storage.dbConnections.closeAllConnections();
        });

        return storage;
    }

    /**
     * Returns the global configuration instance. It will respect the environment variables.
     * As opposed to this method, we can also get the SDK instance configuration via `sdk.config` property.
     *
     * @return {Configuration}
     */
    static getGlobalConfig() {
        if (!Configuration.globalConfig) {
            Configuration.globalConfig = new Configuration();
        }

        return Configuration.globalConfig;
    }
}
